### ISSUE TITLE
Bad encrypted message, in 1-on-1 chat, and also in never-used Note To Self 

### ISSUE THREAD

**Issue by five-c-d** [signalapp/Signal-Android#8663]
- [x] I have searched open and closed issues for duplicates
- [x] I am submitting a bug report for existing functionality that does not work as intended
- [x] I have read https://github.com/signalapp/Signal-Android/wiki/Submitting-useful-bug-reports
- [x] This isn't a feature request or a discussion topic

----------------------------------------

### Bug description
signal4android processes *batches* of incoming messages from the signal-server delivery-queue too noisily, and painfully inefficiently, if it has been offline for any appreciable amount of time, and is brought back online with some purpose in mind (other than the quasi-purpose of "letting signal4android catch up on messages"). 

Sometimes, this story ends up very poorly, in a string of hundreds and hundreds of Bad Encrypted Message errors (BEM henceforth).  Please find below, one such adventure. 

### Steps to reproduce
-  first, be having a signal4desktop#A conversation with another person on signal4desktop#B which is linked to their android#B 
-  second, decide that our textual-conversation will go more smoothly if we can cryptocall on speakerphone whilst we simultaneously continue typing on our signal4desktop instances 
-  third, connect android#A to the internet, after roughly 48 hours of being in airplane-mode ... with intent to make that cryptocall firmly in mind.  
-  all throughout this saga (outlined below), android#A is dinging&buzzing as it tries to catch up on old news (signal4desktop conversation it missed because of airplane-mode) 

<details> <summary> gory details </summary> <p> 

-  attempted to initiate cryptocall ~~14:21 from android#A to android#B , carefully dodging the constant popups by tapping the call-icon at just the right moment 
-  attempted call 14:22 from B to A, shows as "missed call" on A
-  attempted call 14:22 from A to B
-  attempted call 14:23 from B to A, shows as "missed call" on A
-  comment by A:  dialing, but never makes it to ringing (sent from desktop#A and received by android#A) 
-  attempted call 14:25 from B to A, shows as "missed call" on A
-  B presses hangup, but android#A does not figure that out
-  attempted call 14:25 from A to B, but what appears on A is the **incoming** call screen ... no audible ringing ... says 'INCOMING CALL' in the cyan middle-bar ... swipe-up does not work (B already hung up) 
-  attempted call 14:28 from B to A, shows as "missed call" on A
-  attempted call 14:28 from B to A, shows as "missed call" on A
-  attempted call 14:28 from A to B
-  first Bad Encrypted Message appears on A, dated sunday 11:43 , then 3 successes stamped 11:44, then another BEM at 11:44, +8 good, 1 BEM, 1 good, 1 BEM, 5 good, 1 BEM, and so on like that.  BEM are always the incoming ones (from signal4desktop#B running windows to just-catching-up android#A device which has been offline), outgoing messages from signal4desktop#A do not appear as BEM on android#A that I could tell.  
-  attempted call 14:30 from A to B ... android#B sees the incoming-call screen, and can attempt to swipe-up / swipe-down the button, but "nothing happens when I try" to do those actions ... android#A meanwhile always displays 'DIALING' never ringing 
-  at this point every message that android#A receives during the sync-and-download attempt begins showing as BEM ... wall-time is 14:31 monday, but android#A is still working on sunday 11:54 stamped chat-bubbles.  after roughly ~150 BEM in a row (dinging android#A once every couple of seconds despite me having signalapp open to the chat in question), over the course of ~3 mins wall-clock elapsed, android#A gets to sunday 13:07 stamped messages.  
-  sent some text-messages from android#A for the first time, at 14:34, which did not work properly ... shows double-unfilled-checkmarks on android#A (i.e. delivered no read-receipt ever properly processed) whereas over on desktop#A shows double-FILLED-checkmarks (i.e. the sync worked and the message went out to android#B plus desktop#B and one or both of them sent the read-receipt which desktop#A received and properly processed) 
-  attempted call 14:35 from B to A, shows as "missed call" on A
-  attempted call 14:35 from B to A, shows as "missed call" on A
-  after waiting another couple minutes, as of 14:36 android#A was still throwing BEM from message-traffic the previous day 
-  after another five minutes of steady monotonous ding&buzz BEM BEM BEM, decided to try and reset encrypted session on android#A at 14:41  
-  enduser of android#B and desktop#B reported that **no** error messages of any sort were appearing on their devices, all during the preceding twenty minutes of cryptocall-attempts ... desktop#A was also unaware of any problems at the time 
-  this did not work... and I believe, but am not positive, that after session-reset *zero* error-messages appeared for the enduser with android#B and desktop#B ... they did see the session-reset alert, though.  
-  android#A continued to popup, ding, buzz, and BEM every second much like before... desktop#A continued to be oblivious and able to exchange text-messages with desktop#B plus android#B with no discernable delays 
-  android#A threw a couple of "messge encrypted for non-existent session" errors shortly after it noticed the session had been reset, then went back to "Bad encrypted message" almost immediately ... probably 350 in a row? 
-  three of the incoming messages to android#A, near the end of that long string of BEM ding&buzz events, were timestamped Monday 14:41 and 14:42 ... though I believe they did not RENDER at that wall-clock time  :-)  
-  finally, with timestamp 14:50 for all three, signal4desktop#A displayed three error-messages in their individual chat-bubbles in a row:  Error handling incoming message ... message-details on the first of the three:  Bad MAC. Sent Monday, March 4, 2019 2:50 PM (1551732630655).  Received Monday, March 4, 2019 2:50 PM (1551732633979).  sender:  enduser#B.  sealedSender:  enabled.  
-  at some point around now, closed signal4desktop#A and re-launched it, plus powered off signal4android#A ... which kept dinging even from the OS-is-shutting-down-animation-screen... then re-powered-up android#A.  may have also reset the session a second time, around 14:50, methinks?  
-  There were another couple dozen BEMs, between 14:50 and 14:57, interspersed with "successful" messages from desktop#A or from desktop#B at unpredictable points in wall-clock time 
-  after a **bit** more waiting around, by 14:57 android#A was finally somewhat back to normal.  
-  cryptocall from A to B, was 'DIALLING' for less than two seconds, then 'RINGING' for less than two seconds.  spoke briefly, hung up, captured four debuglogs
-  cryptocall from A to B again at 15:00 worked just fine as well.   

</p> </details> 

<br />

**Actual result:** 

*  total downtime was 40 minutes
*  this was android#A attempting to sync roughly 48 hours of chat-history, on the order of 400 to 600 chat-messages, while on wifi
*  aka throughput of around one message every five seconds roughly.  

**Expected result:** 

*  the expected behavior, is that signal4android needs to hurry up and help me accomplish what I want, which is to be able to make a cryptocall or send an encrypted message or whatever 
*  it can tell, by contacting signal4desktop, that all the messages is it about to download, are things that I've already viewed on another device linked to it 
*  so the key is therefore to achieve a high-performance pre-processing phase, where that big old somewhat-pointless download can double-ratchet along *quickly* to the point where signal4android is useful/usable ASAP 

<details> <summary> optimization advice </summary> <p> 

*  android#A needs to open a websocket to signal-server.  
*  android#A needs to download every single message in a stream of packets, quickly and without needless roundtrips and without interruption, emptying the delivery-queue like it was a streaming-video containing 500 movie-frames, and NOT like it was 500 little individual files that each required their own HTTP request and their own HTTP response with pauses for processing before and after each network-operation 
*  android#A needs to decrypt the minimum possible from those streaming-downloads, so as to strip off the sealed-sender protections and figure out which chat-history each new message-bubble goes in, and the metadata-timestamp
*  android#A should not perform ANY notification-popups during this entire process:  no ding, no vibrate, no screen-pop 
*  android#A should not perform ANY disk-writes during this process, unless physical RAM is so limited that it needs to dump some of the incoming stream of bytes to an (encrypted) binary scratchfile temporarily 
*  android#A should not perform any unnecessary SQLCipher queries (especially not write-queries) during this entire process.  
*  android#A has the goal of getting to where it is operational again, and the enduser can make a cryptocall or send a crypto-text, which is WHY the enduser just opened it up, after all 
*  android#A should defer everything that can be deferred, to get to that point in time.  
*  there should be no messing around with read-receipts, is-typing, generating sync-messages to send out to other linked-devices, pulling attachments from S3, uploading addressbook-info, checking whether anybody has altered their signal-profile-avatar, checking whether any old disappearing-messages have gone past their expiration-date and need secure_delete, ANYTHING but the priority of churning through the delivery-queue as quickly as possible so as to be able to **live in the now**  ;-)  
*  android#A is *then* free, only then, in the background and without interrupting the enduser's ongoing activity (making a cryptocall or sending textual messages or whatever), to begin rendering message-bodies into the correct on-screen locations, begin storing fully-decrypted stuff into SQLCipher locally-encrypted, and generally getting all the eventually-needs-to-be-done things accomplished, which were brutally skipped during the get-to-operational pre-processing phase outlined above.  
*  android#A will, as part of the post-processing effort, eventually get to the point where it is fully up to date and completely synchronized and has no partly-decrypted envelopes and scratchfiles and other such things laying around.  
*  this entire approach is considerably more complex than the current approach, which assumes that nobody will ever leave their phone offline for 48 hours and nobody will ever receive 480 messages in that timespan
*  but the advantage to this approach, is that instead of processing messages at the rate of one every five seconds on average, it ought to be possible to pre-process messages at the rate of *five per second*, without even trying very hard to optimize anything, just merely avoiding stalling the download-pipeline, avoiding disk-writes that must be flushed before anything can continue, and avoiding CPU cycles which can be deferred until later for the lazy-compute background post-processing phase.  

</p> </details> 

<br />

*  Therefore, instead of it taking FORTY minutes to get operational, dinging and buzzing every second along the way, it might only take 90 seconds to get operational... with a progress-bar and an ETA of "pre-processing #100 of 480 messages" ... and in blissful silence 
*  At the end of the pre-processing phase, there can be **one** ding and **one** notification-popup:  "you have 456 new messages in 23 conversations" 

### Screenshots

Imagine seventeen android-display screenfuls of [Bad encrypted message] in a row.  Wish that I was kidding  :-)   Besides in the 1-on-1 chat where I was most active during that 48-hour timespan, there was also a large number -- many dozen -- BEM chat-bubbles in the NoteToSelf on android#A, which I have never actually used.  Unclear why.  Both the 1-on-1 chat and the NoteToSelf are working fine now.  

### Device info

**Device:** Alcatel A574BL
**Android version:** 7.1.1 
**Signal version:** 4.34.8 

### Link to debug logs

*  android#A == https://debuglogs.org/cb4e503f1fbb6cfd85735b8c33c7d261cb9cac6c59c4ae6f2d927ab97bf7198b
*  desktop#A == https://debuglogs.org/9f25d474f228ec3ebfe32982288057becdbeb226a3e531b820c0a71350cd473e
*  desktop#B == https://debuglogs.org/3c73cbe93c6d91792be32c0665698606ef80aebb564a7b671a3ee7ef2d94c8fd
*  android#B == https://debuglogs.org/be9e208b7534761e2fc45472e704d2e44a2cde67e8094fe106d839954eab80d4

Unfortunately, the vast outpouring of BEM and friends, almost certainly means that the root cause of the trouble cannot be traced, because the debuglogs are empty of the portion of the 40-minute timespan that would have potentially revealed what went wrong.  However, possibly by correlating desktop#A with android#A logfiles, it will become clear what occurred in the early phases of android#A's difficulties.  Note that the chat-bubble ordering of android#A is *completely* out of whack (messages from the three days all getting interspersed and intermixed) but that desktop#A was powered-on and internet-connected at all times so it has a properly-ordered transcript of message-bodies for the most part. 

### Potentially related 

* https://github.com/signalapp/Signal-Android/issues/7465#issuecomment-378147544 , Bad old messages after database upgrade 
* https://github.com/signalapp/Signal-Desktop/issues/3061 , desktop Spams Contacts After Not Being Used for Over A Month
* https://github.com/signalapp/Signal-Desktop/issues/3050 , desktop sends message with text "Bad encrypted message" multiple times to contacts
* https://github.com/signalapp/Signal-Desktop/issues/1751 , Messages sent from desktop appear as a 'Bad encrypted message' in mobile
* https://github.com/signalapp/Signal-Desktop/issues/2912 , Error Handling Incoming Message - Bad MAC 
* https://github.com/signalapp/Signal-Android/issues/7154 , Bad encrypted message from myself [however, this was long before Note To Self ... note that I had used the unofficial dial-my-own number trick before, but never utilized the official new NoteToSelf feature yet on android#A ... so why it was getting BEM in that is mysterious to me, perhaps those are sync-messages from desktop#A being misinterpreted as NoteToSelf?] 
* <a href="https://github.com/search?o=desc&q=org%3Asignalapp+%22bad+encrypted+message%22&s=created&type=Issues">et cetera </a>

**Comment by atoav** [comment]
Had the same happen to me, with the same symptoms (I use both a Galaxy S7 and a Windows Desktop Client)

**Comment by 6lancmange** [comment]
Yep, I think I'm having the same problem. My Android phone was off for several days and I was only using Signal on desktop, so there had to be hundreds of messages. I suspected there would be a problem as soon as I get my phone back and turn on Signal. When it started syncing, all off the messages are just "Bad encrypted message", and I got many Notes to Self – most of them just outgoing messages that I sent on desktop. There was a problem on the desktop with a message I sent from Android, but I think it was fixed after I reset the session.

Is there something to do to get the messages on Android, or are they forever lost, only able to be seen on desktop?

Here's the debug logs:
Android – https://debuglogs.org/5ec3b1dbb3b8b55b6dd52c4fefc756ff6771d1ca490bccb2ac1510051a512320
Desktop (Ubuntu 18.04.3 LTS) – 
https://debuglogs.org/160c301b5c3fba8a0d416256e179d5f640b52042c1f3c18fefd794226b8ef645
